pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java

```diff
@@ -67,16 +72,28 @@ public boolean isNoopTypeCheck() {
   public abstract TypeNode initWriteSlotNode(int slot);
 
   /**
-   * Checks if `value` conforms to this type. If so, returns normally. Otherwise, throws a
-   * `VmTypeMismatchException`.
+   * Checks if {@code value} conforms to this type, and possibly casts it in the case of {@link
+   * MappingTypeNode} or {@link ListingTypeNode}.
    */
-  public abstract void execute(VirtualFrame frame, Object value);
+  public abstract Object execute(VirtualFrame frame, Object value);
```

Comment:
This is one of the biggest changes. Executing a typecheck will return a type-casted value.

For most types, this simply returns `value` if the value conforms to the type (e.g. `Int`). For `Listing` and `Mapping`, this returns a surrogate listing/mapping object. Other parameterized types may also return a new object if the parameterized type is `Listing` or `Mapping` (e.g. `List<Listing<Int>>`).